homeassistant/helpers/state.py

```diff
@@ -132,6 +99,103 @@ def get_changed_since(states: Iterable[State],
             if state.last_updated >= utc_point_in_time]
 
 
+@bind_hass
+def async_map_state_to_calls(hass: HomeAssistantType,
+                             domain: str,
+                             state: State,
+                             state_to_service: Dict[str, str],
+                             services: Dict[str,
+                                            voluptuous.schema_builder.Schema]
+                             ) -> List[ServiceCall]:
+        """
+        Maps a state object into service calls to reproduce the given state.
+
+        Function will make two passes. First it will create a call to reproduce
+        the state.state, passing along any allowed attributes to the state
+        setting service. As a second pass it will map remaining attributes
+        to services
+        """
+
+        calls = []
+
+        # expanded attribute list containing entity id
+        expanded_attributes = dict(state.attributes)
+        expanded_attributes[ATTR_ENTITY_ID] = state.entity_id
+
+        def add_if_complete(service: str) -> None:
+            """Add a service call with the attributes needed."""
+            schema = voluptuous.schema_builder.Schema(
```

Comment:
It's more flexible if we define the schema in each component, and accept the schema in the helper.

All voluptuous names are available under `voluptuous.`, so `voluptuous.Schema` is enough. We usually import `voluptuous as vol`.

Reply:
It's like that already. It's just that i need to modify the schema to drop extra data and not bail out on extra attributes.

Reply:
I mean that instead of passing a dict that maps states to services we pass a schema.

Reply:
Another point is that it's more flexible if we keep the knowledge about how he schema is made in each component, and just let the helper validate the state against the schema and if it passes, store the service call.

Reply:
we still need states to service and whitelisted services. some services take no parameter or toggle states. for example light.toggle. it takes optional transition. This services can't be allowed (with default schema anyway).

Also if it is allowed with required transition attribute, its one of those that require consumption of attributes. Otherwise turn_on will turn on, with a transition for example, the the toggle service gets called with the remaining attribute.

Reply:
I'm not saying we should use the default schema. I'm saying we should define each schema in each component. We can build off the service schema when available. If not available, the component makes one from scratch.

Reply:
I'm don't understand the turn on + transition + toggle example. If we turn on, we don't toggle, right?

Reply:
I'm not sure i follow. We have two things in play:
1: we need to map "state: on/off/heat" to a service taking no parameters.
2: we need to map additional attributes to additional services.

For 1, i don't think we can do without a dict? they cant be differentiated by schema.
For 2. the schema is already passed from component.

Reply:
A schema can use a dict to validate.

Reply:
We need to add the allowance of extra attributes somewhere. Otherwise the code ends up only able to call a single service as the original.

So the schemas registered for the service in the registry are never ever okey as is.

Reply:
Yep, it's fine to make new schemas based on the existing service schemas. But the helper shouldn't do that. That's all I'm saying.